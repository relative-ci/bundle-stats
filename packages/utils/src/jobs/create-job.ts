import merge from 'lodash/merge';

import { SOURCE_PATHS } from '../config';
import { createSummary } from './create-summary';
import * as webpack from '../webpack';
import * as lighthouse from '../lighthouse';
import * as browsertime from '../browsertime';
import { JobData, JobMetricsSource, SourceData } from '../constants';

interface SourceResult {
  meta: {};
  insights: {};
  metrics: JobMetricsSource;
}

interface SourceModule {
  SUMMARY_METRIC_PATHS: Array<string>;
  extract: (rawData: any, balien: any) => any;
}

const SOURCE_MODULES: Record<string, SourceModule> = { webpack, lighthouse, browsertime };

export const createJob = (source: SourceData, baseline?: JobData): JobData =>
  SOURCE_PATHS.reduce((agg, sourcePath) => {
    const rawData = source[sourcePath];

    if (!rawData) {
      return agg;
    }

    const sourceModule = SOURCE_MODULES[sourcePath];

    if (!sourceModule) {
      return agg;
    }

    const extractedData: SourceResult = sourceModule.extract(rawData, baseline);

    const summary = createSummary(
      SOURCE_MODULES[sourcePath].SUMMARY_METRIC_PATHS,
      baseline?.metrics?.[sourcePath],
      extractedData?.metrics,
    );

    return merge({}, agg, {
      meta: {
        [sourcePath]: extractedData.meta,
      },
      insights: {
        [sourcePath]: extractedData.insights,
      },
      summary: {
        [sourcePath]: summary,
      },
      metrics: {
        [sourcePath]: extractedData.metrics,
      },
      rawData: {
        [sourcePath]: rawData,
      },
    });
  }, {});
