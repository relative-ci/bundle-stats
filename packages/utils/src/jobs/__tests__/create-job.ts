import webpackStatsCurrentExtracted from '../../../__fixtures__/webpack-stats-1.extracted';
import webpackStatsBaselineExtracted from '../../../__fixtures__/webpack-stats-2.extracted';

import { createJob } from '../create-job';

describe('createJob', () => {
  test('no baseline', () => {
    const actual = createJob({ webpack: webpackStatsCurrentExtracted });
    expect(actual).toMatchSnapshot();
  });

  test('with baseline', () => {
    const baselineJob = createJob({ webpack: webpackStatsBaselineExtracted });
    const actual = createJob({ webpack: webpackStatsCurrentExtracted }, baselineJob);
    expect(actual).toMatchSnapshot();
  });
});
