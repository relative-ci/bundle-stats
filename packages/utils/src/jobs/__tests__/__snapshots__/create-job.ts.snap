// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createJob no baseline 1`] = `
{
  "insights": {
    "webpack": {
      "assetsSizeTotal": {
        "data": {
          "info": {
            "delta": 63000,
            "deltaPercentage": 100,
            "deltaType": "HIGH_NEGATIVE",
            "displayDelta": "+61.52KiB",
            "displayDeltaPercentage": "+100%",
            "displayValue": "61.52KiB",
            "value": 63000,
          },
          "md": "*Bundle Size* — *61.52KiB* (*+100%*).",
          "text": "Bundle Size — 61.52KiB (+100%).",
        },
        "type": "INFO",
      },
      "duplicatePackages": {
        "data": {
          "package-c": [
            "package-a:package-c",
            "package-c",
          ],
        },
        "type": "WARNING",
      },
    },
  },
  "meta": {
    "webpack": {
      "builtAt": "2019-01-01T00:00:00.000Z",
      "chunks": [
        {
          "id": "1",
          "name": "main",
        },
      ],
      "hash": "abcd1234",
    },
  },
  "metrics": {
    "webpack": {
      "assetCount": {
        "value": 4,
      },
      "assets": {
        "index.html": {
          "isChunk": false,
          "isEntry": false,
          "isInitial": false,
          "name": "index.html",
          "value": 2000,
        },
        "logo.png": {
          "isChunk": false,
          "isEntry": false,
          "isInitial": false,
          "name": "logo.a3b4c5d.png",
          "value": 1000,
        },
        "main.css": {
          "chunkId": "1",
          "isChunk": true,
          "isEntry": true,
          "isInitial": true,
          "name": "main.a1b2c3d.css",
          "value": 10000,
        },
        "main.js": {
          "chunkId": "1",
          "isChunk": true,
          "isEntry": true,
          "isInitial": true,
          "name": "main.a2b3c4d.js",
          "value": 50000,
        },
      },
      "cacheInvalidation": {
        "value": 0,
      },
      "chunkCount": {
        "value": 2,
      },
      "duplicateCode": {
        "value": 0,
      },
      "duplicateModulesCount": {
        "value": 0,
      },
      "duplicatePackagesCount": {
        "value": 1,
      },
      "moduleCount": {
        "value": 6,
      },
      "modules": {
        "module-a": {
          "chunkIds": [
            "1",
          ],
          "duplicated": false,
          "name": "module-a",
          "value": 1000,
        },
        "module-b": {
          "chunkIds": [
            "1",
          ],
          "duplicated": false,
          "name": "module-b",
          "value": 2000,
        },
        "node_modules/package-a/index.js": {
          "chunkIds": [
            "1",
          ],
          "duplicated": false,
          "name": "node_modules/package-a/index.js",
          "value": 1000,
        },
        "node_modules/package-a/node_modules/package-c/index.js": {
          "chunkIds": [
            "1",
          ],
          "duplicated": false,
          "name": "node_modules/package-a/node_modules/package-c/index.js",
          "value": 1000,
        },
        "node_modules/package-b/index.js": {
          "chunkIds": [
            "1",
          ],
          "duplicated": false,
          "name": "node_modules/package-b/index.js",
          "value": 1000,
        },
        "node_modules/package-c/index.js": {
          "chunkIds": [
            "1",
          ],
          "duplicated": false,
          "name": "node_modules/package-c/index.js",
          "value": 1000,
        },
      },
      "packageCount": {
        "value": 4,
      },
      "packages": {
        "package-a": {
          "name": "package-a",
          "path": "node_modules/package-a",
          "value": 1000,
        },
        "package-a:package-c": {
          "name": "package-c",
          "path": "node_modules/package-a/node_modules/package-c",
          "value": 1000,
        },
        "package-b": {
          "name": "package-b",
          "path": "node_modules/package-b",
          "value": 1000,
        },
        "package-c": {
          "name": "package-c",
          "path": "node_modules/package-c",
          "value": 1000,
        },
      },
      "sizes": {
        "totalSizeByTypeCSS": {
          "value": 10000,
        },
        "totalSizeByTypeFONT": {
          "value": 0,
        },
        "totalSizeByTypeHTML": {
          "value": 2000,
        },
        "totalSizeByTypeIMG": {
          "value": 1000,
        },
        "totalSizeByTypeJS": {
          "value": 50000,
        },
        "totalSizeByTypeMEDIA": {
          "value": 0,
        },
        "totalSizeByTypeOTHER": {
          "value": 0,
        },
      },
      "totalInitialSizeCSS": {
        "value": 10000,
      },
      "totalInitialSizeJS": {
        "value": 50000,
      },
      "totalSizeByTypeALL": {
        "value": 63000,
      },
    },
  },
  "rawData": {
    "webpack": {
      "assets": [
        {
          "name": "main.a1b2c3d.css",
          "size": 10000,
        },
        {
          "name": "main.a2b3c4d.js",
          "size": 50000,
        },
        {
          "name": "logo.a3b4c5d.png",
          "size": 1000,
        },
        {
          "name": "index.html",
          "size": 2000,
        },
      ],
      "builtAt": 1546300800000,
      "chunks": [
        {
          "entry": true,
          "files": [
            "main.a1b2c3d.css",
            "main.a2b3c4d.js",
          ],
          "id": 1,
          "initial": true,
          "names": [
            "main",
          ],
        },
      ],
      "hash": "abcd1234",
      "modules": [
        {
          "chunks": [
            1,
          ],
          "name": "module-a",
          "size": 1000,
        },
        {
          "chunks": [
            1,
          ],
          "name": "module-b",
          "size": 2000,
        },
        {
          "chunks": [],
          "name": "module-c",
          "size": 1000,
        },
        {
          "chunks": [
            1,
          ],
          "name": "node_modules/package-a/index.js",
          "size": 1000,
        },
        {
          "chunks": [
            1,
          ],
          "name": "node_modules/package-a/node_modules/package-c/index.js",
          "size": 1000,
        },
        {
          "chunks": [
            1,
          ],
          "name": "node_modules/package-b/index.js",
          "size": 1000,
        },
        {
          "chunks": [
            1,
          ],
          "name": "node_modules/package-c/index.js",
          "size": 1000,
        },
      ],
    },
  },
  "summary": {
    "webpack": {
      "assetCount": {
        "baseline": 0,
        "current": 4,
      },
      "cacheInvalidation": {
        "baseline": 0,
        "current": 0,
      },
      "chunkCount": {
        "baseline": 0,
        "current": 2,
      },
      "duplicateCode": {
        "baseline": 0,
        "current": 0,
      },
      "duplicateModulesCount": {
        "baseline": 0,
        "current": 0,
      },
      "duplicatePackagesCount": {
        "baseline": 0,
        "current": 1,
      },
      "moduleCount": {
        "baseline": 0,
        "current": 6,
      },
      "packageCount": {
        "baseline": 0,
        "current": 4,
      },
      "totalInitialSizeCSS": {
        "baseline": 0,
        "current": 10000,
      },
      "totalInitialSizeJS": {
        "baseline": 0,
        "current": 50000,
      },
      "totalSizeByTypeALL": {
        "baseline": 0,
        "current": 63000,
      },
    },
  },
}
`;

exports[`createJob with baseline 1`] = `
{
  "insights": {
    "webpack": {
      "assetsSizeTotal": {
        "data": {
          "info": {
            "delta": 0,
            "deltaPercentage": 0,
            "deltaType": "NO_CHANGE",
            "displayDelta": "0B",
            "displayDeltaPercentage": "0%",
            "displayValue": "61.52KiB",
            "value": 63000,
          },
          "md": "*Bundle Size* — *61.52KiB* (*0%*).",
          "text": "Bundle Size — 61.52KiB (0%).",
        },
        "type": "INFO",
      },
      "duplicatePackages": {
        "data": {
          "package-c": [
            "package-a:package-c",
            "package-c",
          ],
        },
        "type": "WARNING",
      },
    },
  },
  "meta": {
    "webpack": {
      "builtAt": "2019-01-01T00:00:00.000Z",
      "chunks": [
        {
          "id": "1",
          "name": "main",
        },
      ],
      "hash": "abcd1234",
    },
  },
  "metrics": {
    "webpack": {
      "assetCount": {
        "value": 4,
      },
      "assets": {
        "index.html": {
          "isChunk": false,
          "isEntry": false,
          "isInitial": false,
          "name": "index.html",
          "value": 2000,
        },
        "logo.png": {
          "isChunk": false,
          "isEntry": false,
          "isInitial": false,
          "name": "logo.a3b4c5d.png",
          "value": 1000,
        },
        "main.css": {
          "chunkId": "1",
          "isChunk": true,
          "isEntry": true,
          "isInitial": true,
          "name": "main.a1b2c3d.css",
          "value": 10000,
        },
        "main.js": {
          "chunkId": "1",
          "isChunk": true,
          "isEntry": true,
          "isInitial": true,
          "name": "main.a2b3c4d.js",
          "value": 50000,
        },
      },
      "cacheInvalidation": {
        "value": 95.24,
      },
      "chunkCount": {
        "value": 2,
      },
      "duplicateCode": {
        "value": 0,
      },
      "duplicateModulesCount": {
        "value": 0,
      },
      "duplicatePackagesCount": {
        "value": 1,
      },
      "moduleCount": {
        "value": 6,
      },
      "modules": {
        "module-a": {
          "chunkIds": [
            "1",
          ],
          "duplicated": false,
          "name": "module-a",
          "value": 1000,
        },
        "module-b": {
          "chunkIds": [
            "1",
          ],
          "duplicated": false,
          "name": "module-b",
          "value": 2000,
        },
        "node_modules/package-a/index.js": {
          "chunkIds": [
            "1",
          ],
          "duplicated": false,
          "name": "node_modules/package-a/index.js",
          "value": 1000,
        },
        "node_modules/package-a/node_modules/package-c/index.js": {
          "chunkIds": [
            "1",
          ],
          "duplicated": false,
          "name": "node_modules/package-a/node_modules/package-c/index.js",
          "value": 1000,
        },
        "node_modules/package-b/index.js": {
          "chunkIds": [
            "1",
          ],
          "duplicated": false,
          "name": "node_modules/package-b/index.js",
          "value": 1000,
        },
        "node_modules/package-c/index.js": {
          "chunkIds": [
            "1",
          ],
          "duplicated": false,
          "name": "node_modules/package-c/index.js",
          "value": 1000,
        },
      },
      "packageCount": {
        "value": 4,
      },
      "packages": {
        "package-a": {
          "name": "package-a",
          "path": "node_modules/package-a",
          "value": 1000,
        },
        "package-a:package-c": {
          "name": "package-c",
          "path": "node_modules/package-a/node_modules/package-c",
          "value": 1000,
        },
        "package-b": {
          "name": "package-b",
          "path": "node_modules/package-b",
          "value": 1000,
        },
        "package-c": {
          "name": "package-c",
          "path": "node_modules/package-c",
          "value": 1000,
        },
      },
      "sizes": {
        "totalSizeByTypeCSS": {
          "value": 10000,
        },
        "totalSizeByTypeFONT": {
          "value": 0,
        },
        "totalSizeByTypeHTML": {
          "value": 2000,
        },
        "totalSizeByTypeIMG": {
          "value": 1000,
        },
        "totalSizeByTypeJS": {
          "value": 50000,
        },
        "totalSizeByTypeMEDIA": {
          "value": 0,
        },
        "totalSizeByTypeOTHER": {
          "value": 0,
        },
      },
      "totalInitialSizeCSS": {
        "value": 10000,
      },
      "totalInitialSizeJS": {
        "value": 50000,
      },
      "totalSizeByTypeALL": {
        "value": 63000,
      },
    },
  },
  "rawData": {
    "webpack": {
      "assets": [
        {
          "name": "main.a1b2c3d.css",
          "size": 10000,
        },
        {
          "name": "main.a2b3c4d.js",
          "size": 50000,
        },
        {
          "name": "logo.a3b4c5d.png",
          "size": 1000,
        },
        {
          "name": "index.html",
          "size": 2000,
        },
      ],
      "builtAt": 1546300800000,
      "chunks": [
        {
          "entry": true,
          "files": [
            "main.a1b2c3d.css",
            "main.a2b3c4d.js",
          ],
          "id": 1,
          "initial": true,
          "names": [
            "main",
          ],
        },
      ],
      "hash": "abcd1234",
      "modules": [
        {
          "chunks": [
            1,
          ],
          "name": "module-a",
          "size": 1000,
        },
        {
          "chunks": [
            1,
          ],
          "name": "module-b",
          "size": 2000,
        },
        {
          "chunks": [],
          "name": "module-c",
          "size": 1000,
        },
        {
          "chunks": [
            1,
          ],
          "name": "node_modules/package-a/index.js",
          "size": 1000,
        },
        {
          "chunks": [
            1,
          ],
          "name": "node_modules/package-a/node_modules/package-c/index.js",
          "size": 1000,
        },
        {
          "chunks": [
            1,
          ],
          "name": "node_modules/package-b/index.js",
          "size": 1000,
        },
        {
          "chunks": [
            1,
          ],
          "name": "node_modules/package-c/index.js",
          "size": 1000,
        },
      ],
    },
  },
  "summary": {
    "webpack": {
      "assetCount": {
        "baseline": 4,
        "current": 4,
      },
      "cacheInvalidation": {
        "baseline": 0,
        "current": 95.24,
      },
      "chunkCount": {
        "baseline": 2,
        "current": 2,
      },
      "duplicateCode": {
        "baseline": 0,
        "current": 0,
      },
      "duplicateModulesCount": {
        "baseline": 0,
        "current": 0,
      },
      "duplicatePackagesCount": {
        "baseline": 0,
        "current": 1,
      },
      "moduleCount": {
        "baseline": 5,
        "current": 6,
      },
      "packageCount": {
        "baseline": 3,
        "current": 4,
      },
      "totalInitialSizeCSS": {
        "baseline": 11000,
        "current": 10000,
      },
      "totalInitialSizeJS": {
        "baseline": 49000,
        "current": 50000,
      },
      "totalSizeByTypeALL": {
        "baseline": 63000,
        "current": 63000,
      },
    },
  },
}
`;
